jetty-maven-plugin/src/main/java/org/eclipse/jetty/maven/plugin/JettyStopMojo.java

```diff
@@ -72,38 +74,132 @@ public void execute() throws MojoExecutionException, MojoFailureException
+                        getLog().info("Couldn't verify server as stopped, received " + response);
+                }
+                else
+                {
+                    //wait for pid to stop
+                    getLog().info("Waiting " + stopWait + " seconds for jetty " + pid + " to stop");
+                    send(stopKey + "\r\n" + command + "\r\n", false, 0);
+                    Optional<ProcessHandle> optional = ProcessHandle.of(pid);
+                    ProcessHandle handle = optional.orElse(null);
+                    if (handle != null)
+                    {
+                        try
+                        {
+                            CompletableFuture<ProcessHandle> future = handle.onExit();
+                            if (handle.isAlive())
+                            {
+                                handle = future.get(stopWait, TimeUnit.SECONDS);
+                            }
+
+                            if (handle.isAlive())
+                                getLog().info("Couldn't verify server process stop");
+                            else
+                                getLog().info("Server process stopped");
+                        }
+                        catch (Throwable e)
+                        {
+                            getLog().error(e);
+                        }
                     }
                 }
             }
+            catch (ConnectException e)
+            {
+                getLog().info("Jetty not running!");
+            }
+            catch (Exception e)
+            {
+                getLog().error(e);
+            }
         }
-        catch (ConnectException e)
+        else
         {
-            getLog().info("Jetty not running!");
+            //send the stop command but don't wait to verify the stop
+            getLog().info("Stopping jetty");
+            try
+            {
+                send(stopKey + "\r\n" + command + "\r\n", false, 0);
+            }
+            catch (ConnectException e)
+            {
+                getLog().info("Jetty not running!");
+            }
+            catch (Exception e)
+            {
+                getLog().error(e);
+            }
         }
-        catch (Exception e)
+    }
+    
+    /**
+     * Send a command to a jetty process, optionally waiting for a response.
+     * 
+     * @param command the command to send
+     * @param expectResponse if true, we will wait for a response
+     * @param wait length of time in sec to wait for a response
+     * @return the response, if any, to the command
+     * @throws Exception
+     */
+    private String send(String command, boolean expectResponse, int wait)
+        throws Exception
+    {
+        String response = null;
+        try (Socket s = new Socket(InetAddress.getByName("127.0.0.1"), stopPort); OutputStream out = s.getOutputStream();)
         {
-            getLog().error(e);
+            out.write(command.getBytes());
+            out.flush();
```

Comment:
No, TCP is not very friendly with open/write/close.
You should always read to have a confirmation that the other peer received the data (even if it's only reading -1).
The other machine could be overloaded, so SYN+PSH+FIN may be received by the recipient at OS level (but not yet at application level), but because the machine is overloaded timeouts may happen, so the OS decides to RST back to the sender, which has already closed. 
The net result is that the sender does not know that something went wrong, the recipient never received the data at the application level, and everything went into /dev/null.

Reply:
If the sender is interested in getting confirmation from the remote jetty, then they configure a non-zero `stopWait` timeout, in which case we set the socket soTimeout and read from it to get the "Shutting down" message.  Alternatively, if we have been able to obtain the pid, we send the message and then wait up to `stopWait` seconds on the process itself to stop. So in either case, there will be a wait for either a response message, or the process to die.  If no wait is configured, then we don't wait for any response, we just send the message and exit.

Reply:
You are assuming that the message you send reaches the other end. I'm saying that it may not.
Now, if you care whether the message reaches the other end, you should read, at least a -1.
If you don't care whether the message reaches the other end, then it's fine.

Reply:
So @sbordet how long does the sender need to wait to read from the socket to ensure that the message was sent? The behaviour of the `ShutdownMonitor` is to stop jetty _before_ replying with "Stopped" and exiting. In the case where the caller has _not_ configured a `stopWait` time they do _not_ want to wait until the socket is closed etc.